tests/unit/s2n_async_offload_cb_test.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "error/s2n_errno.h"
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_async_offload.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_security_policies.h"
+#include "utils/s2n_safety.h"
+
+struct s2n_async_offload_cb_test {
+    unsigned invoke_perform : 1;
+    int result;
+
+    int invoked_count;
+    struct s2n_async_op *op;
+};
+
+int s2n_async_offload_test_callback(struct s2n_connection *conn, struct s2n_async_op *op, void *ctx)
+{
+    EXPECT_NOT_NULL(op);
+
+    struct s2n_async_offload_cb_test *data = (struct s2n_async_offload_cb_test *) ctx;
+    data->invoked_count += 1;
+    data->op = op;
+
+    if (data->invoke_perform) {
+        EXPECT_SUCCESS(s2n_async_op_perform(op));
+    }
+
+    return data->result;
+}
+
+static int s2n_test_handshake_async(struct s2n_connection *server_conn, struct s2n_connection *client_conn,
+        struct s2n_async_offload_cb_test *data)
+{
+    while (true) {
+        int ret_val = s2n_negotiate_test_server_and_client(server_conn, client_conn);
+
+        if (s2n_errno == S2N_ERR_ASYNC_BLOCKED) {
+            /* Handshake remains blocked as long as op_perform() is not invoked. */
+            for (int i = 0; i < 3; i++) {
+                EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                        S2N_ERR_ASYNC_BLOCKED);
+            }
+
+            EXPECT_SUCCESS(s2n_async_op_perform(data->op));
+            /* Each operation can only be performed once. */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_async_op_perform(data->op), S2N_ERR_INVALID_STATE);
+        } else {
+            return ret_val;
+        }
+    }
+    return 0;
+}
+
+int main(int argc, char *argv[])
+{
+    BEGIN_TEST();
+
+    /* Safety Check */
+    {
+        struct s2n_async_offload_cb_test test_data = { 0 };
+        EXPECT_FAILURE_WITH_ERRNO(s2n_config_set_async_offload_callback(NULL, s2n_async_offload_test_callback, S2N_ASYNC_ALLOW_ALL, &test_data),
+                S2N_ERR_NULL);
```

Comment:
Nit: very long line, wrapping on my monitor. You should manually wrap around 100 so you can control how the wrapping is done. You've got a couple other long lines in this file.